### ISSUE TITLE
Let users specify a Dockerfile with `balena build`

### ISSUE THREAD

**Issue by imrehg** [balena-io/balena-cli#587]
For `docker build` one can specify the `Dockerfile` to build from with the `-f` flag, thus a single directory can contain `Dockerfile`s for multiple deploys. (e.g. `docker build . -f Dockerfile.raspberrypi3`)

Feels like it would be useful if `resin build` can also have a `Dockerfile` specified (instead of just running from the default name).

Also, if `resin build` gets this, probably should expose it to `resin deploy` as well (to run together with the `--build` option).

<img src="https://frontapp.com/assets/img/icons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_gvk3)

**Comment by pimterry** [comment]
We've actually already been discussing this in flowdock after a separate issue looking at this (#575), see https://www.flowdock.com/app/rulemotion/resin-tech/threads/An7HSBVXTvG0OBvlFrXkNEpJorJ.

@CameronDiver what's the status of this?

**Comment by imrehg** [comment]
Thanks @pimterry, missed that thread!

**Comment by pimterry** [comment]
Update: we're pretty busy with other things at the moment (Cameron especially), so while this is still a feature we'd like, I'm not expecting that we'll get around to adding support any time soon.

Do ping here if this is causing you problems though, and we can reevaluate.

**Comment by mechiris** [comment]
Any movement on this? a -f option would be really nice to have.

**Comment by pimterry** [comment]
Not yet I'm afraid @mechiris. Most of the team is currently working hard adding [multicontainer](https://forums.resin.io/t/state-update-multicontainer/1592/8) support throughout Resin. That will soon necessitate some large changes & improvements to `resin build` as well though, and we'll certainly take a look at including support for this as part of that, if we can.

**Comment by pdcastro** [comment]
Vaguely related to #491

**Comment by pdcastro** [comment]
Relevant backend pull request: https://github.com/balena-io-modules/resin-multibuild/pull/38/

**Comment by CameronDiver** [comment]
Work required to integrate the above changes into the CLI:
* Upgrade to multibuild v1.0.0
* When a file cannot be resolved, an error is emitted to the build stream, like so: https://github.com/balena-io-modules/resin-multibuild/pull/38/files#diff-3c8bc921501e1da4e56863d78f60a100R51
* A promise is not returned from `performResolution` anymore, and the resolution information is only available after the stream has been finalised (which normally will be when the build starts)

The composition part of this will automatically be picked up, so just handling these major changes should be enough.

**Comment by pdcastro** [comment]
Related / depends on issue: https://github.com/balena-io/resin-builder/issues/573

**Comment by pdcastro** [comment]
When this issue is resolved, please update this forum thread:
https://forums.balena.io/t/defining-context-in-docker-compose-error-cannot-locate-specified-dockerfile-dockerfile/2894

**Comment by pdcastro** [comment]
Update: CLI version 9.13.0 and later integrated a new major release of [resin-multibuild](https://github.com/balena-io-modules/resin-multibuild/) that allows Dockerfile selection through the docker-compose `dockerfile` instruction, for example:
```
# docker-compose.yml
version: '2'
services:
  test1:
    build:
      context: .
      dockerfile: ./myService/myDockerfile
```
This works for the CLI's `build`, `deploy` and `push` commands, however:
* There's currently a [known issue](https://github.com/balena-io-modules/resin-bundle-resolve/issues/31) that means that this only works when the value of the context instruction is `'.'` (as in the example above), but Dockerfile selection is still possible.
* There's no `-f` command-line option.

**Comment by pdcastro** [comment]
The known issue mentioned in my previous comment has been resolved.
So I understand it's now possible to specify a Dockerfile using the `dockerfile` instruction in a `docker-compose.yml` file (which `balena build` will automatically take into account), so the intent of this GitHub issue is substantially delivered. But there is still no `-f` command-line option to `balena build`, and there could be one for compatibility with `docker build`, so I think this issue remains open.

**Comment by pdcastro** [comment]
CLI version 10.3.0 introduces the `--dockerfile` command-line option for the `balena push` / `build` / `deploy` commands, targeting either the cloud builders or a local device. I works not just with plain Dockerfiles, but also with `.template` and `.[architecture]` Dockerfiles. For example:
```
$ balena deploy myApp -s folder/myApp --build --dockerfile MyDockerfile
$ balena push myApp -s folder/myApp --dockerfile MyDockerfile.armv7hf
$ balena push 192.168.2.3 -s folder/myApp --dockerfile MyDockerfile.template
```
The `push myApp` variant requires cloud builder support which is currently on the staging environment only. I expect it to reach production in the next 48 hours.
